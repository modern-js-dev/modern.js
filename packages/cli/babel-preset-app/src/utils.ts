import fs from 'fs';
import path from 'path';
import semver from 'semver';

export const isBeyondReact17 = (cwd: string) => {
  const pkgInfo = JSON.parse(
    fs.readFileSync(path.join(cwd, 'package.json'), 'utf8'),
  );

  const deps = { ...pkgInfo.devDependencies, ...pkgInfo.dependencies };

  if (typeof deps.react !== 'string') {
    return false;
  }

  return semver.satisfies(semver.minVersion(deps.react)!, '>=17.0.0');
};

// export const isPnpm = (appDirectory: string) => {
//   // pnpm 以及 rush 场景需要编译
//   const currentDir = appDirectory || process.cwd();
//   if (findMonorepoRoot(currentDir)) {
//     return getPkgManager(currentDir) === 'pnpm';
//   } else if (
//     fs.existsSync(path.resolve(currentDir, './pnpm-lock.yaml')) ||
//     fs.existsSync(path.resolve(currentDir, './.rush'))
//   ) {
//     return true;
//   }

//   return false;
// };
