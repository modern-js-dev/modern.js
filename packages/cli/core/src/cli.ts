// 这个文件跟 bin/modern-js.js 基本一样
// 在开发阶段，因为 package.json 的 exports['./bin']['jsnext:source'] 配置
// 了这个文件，所以需要保留, 后续如果找到更好的方式之后会移除这个文件

import { cli } from '.';

const { version } = require('../package.json');

process.env.MODERN_JS_VERSION = version;
if (!process.env.NODE_ENV) {
  process.env.NODE_ENV =
    // eslint-disable-next-line no-nested-ternary
    ['build', 'start', 'deploy'].includes(process.argv[2])
      ? 'production'
      : process.argv[2] === 'test'
      ? 'test'
      : 'development';
}

cli.run(process.argv.slice(2));
