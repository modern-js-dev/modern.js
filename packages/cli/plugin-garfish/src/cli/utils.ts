import type { NormalizedConfig } from '@modern-js/core';

export const makeProvider = () => `
export const provider = function ({basename, dom}) {
  return {
    render({basename, dom, props}) {
      const SubApp = render({ props, basename });
      const node = dom.querySelector('#' + MOUNT_ID) || dom;
      const App = function () {
        return React.createElement(SubApp, props)
      };
      bootstrap(App, node);
    },
    destroy({ dom }) {
      const node = dom.querySelector('#' + MOUNT_ID) || dom;

      if (node) {
        unmountComponentAtNode(node);
      }
    },
    SubModuleComponent: (props) => {
      const SubApp = render({props, basename});

      return createPortal(<SubApp />, dom.querySelector('#' + MOUNT_ID)  || dom);
    },
    jupiter_submodule_app_key: (props) => {
      const SubApp = render({props, basename});

      return createPortal(<SubApp />, dom.querySelector('#' + MOUNT_ID)  || dom);
    }
  }
};

if (typeof __GARFISH_EXPORTS__ !== 'undefined') {
  __GARFISH_EXPORTS__.provider = provider;
}
`;

export const makeRenderFunction = (code: string) => {
  const inGarfishToRender = `
  const { basename, props } = arguments[0] || {};
  let renderByGarfish = false;
  const renderByProvider = !!basename;

  if (IS_BROWSER && window.Garfish && window.Garfish.activeApps && window.Garfish.activeApps.length !== 0) renderByGarfish = true;
  if (IS_BROWSER && window.Garfish && window.Garfish.apps && Object.keys(window.Garfish.apps).length !== 0) renderByGarfish = true;
  if (typeof __GARFISH_EXPORTS__ !== 'undefined') renderByGarfish = true;
  if (renderByGarfish && !renderByProvider) return null;

  function RouterPlugin (routerConfig) {
    if (basename) {
      routerConfig.basename = basename;
      if (routerConfig.supportHtml5History !== false) {
        if (!routerConfig.historyOptions) {
          routerConfig.historyOptions = {
            basename: basename
          };
        } else {
          routerConfig.historyOptions.basename = basename;
        }
      }
    }
    return router(routerConfig);
  }
  `;
  return (
    inGarfishToRender +
    code
      .replace(`router(`, `RouterPlugin(`)
      .replace('IS_BROWSER', `IS_BROWSER && !renderByGarfish`)
  );
};

// support legacy config
export function getRuntimeConfig(config: Partial<NormalizedConfig>) {
  if (config?.runtime?.features) {
    return config?.runtime?.features;
  }
  return config?.runtime || {};
}

// support legacy config
export function setRuntimeConfig(
  config: Partial<NormalizedConfig>,
  key: string,
  value: any,
): undefined {
  if (config?.runtime?.features && config?.runtime?.features[key]) {
    config.runtime.features[key] = value;
    return undefined;
  }
  if (config?.runtime && config?.runtime[key]) {
    config.runtime[key] = value;
    return undefined;
  }
  return undefined;
}
