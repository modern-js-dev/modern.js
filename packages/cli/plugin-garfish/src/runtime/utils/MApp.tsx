import React from 'react';
// eslint-disable-next-line import/no-named-as-default
import Garfish from 'garfish';
import { Manifest, ModulesInfo } from '../useModuleApps';
import { logger, generateSubAppContainerKey } from '../../util';
import { Loadable, MicroProps } from '../loadable';

declare global {
  interface Window {
    modern_manifest?: {
      modules: ModulesInfo;
    };
  }
}

export function generateMApp(
  options: typeof Garfish.options,
  manifest?: Manifest,
) {
  class MApp extends React.Component<MicroProps, any> {
    state: {
      domId: string;
    } = {
      domId: generateSubAppContainerKey(),
    };

    componentDidMount() {
      const { domId } = this.state;
      const { setLoadingState } = this.props;
      const {
        beforeLoad,
        beforeMount,
        errorLoadApp,
        errorMountApp,
        errorUnmountApp,
        ...otherOptions
      } = options;

      // start auto render able
      Garfish.router.setRouterConfig({ listening: true });

      const garfishOptions: typeof Garfish.options = {
        domGetter: `#${domId}`,
        beforeLoad(...args) {
          logger('MApp beforeLoad', args);
          setLoadingState({
            isLoading: true,
            error: null,
          });
          return beforeLoad?.(...args);
        },
        beforeMount(...args) {
          logger('MApp beforeMount', args);
          setLoadingState({
            isLoading: false,
          });
          return beforeMount?.(...args);
        },
        errorLoadApp(error, ...args) {
          logger('MApp errorLoadApp', error, args);
          setLoadingState({
            error,
          });
          return errorLoadApp?.(error, ...args);
        },
        errorMountApp(error, ...args) {
          logger('MApp errorMountApp', error, args);
          setLoadingState({
            error,
          });
          return errorMountApp?.(error, ...args);
        },
        errorUnmountApp(error, ...args) {
          logger('MApp errorUnmountApp', error, args);
          setLoadingState({
            error,
          });
          return errorUnmountApp?.(error, ...args);
        },
        ...otherOptions,
      };

      logger('MApp componentDidMount', {
        garfishRunning: Garfish.running,
        garfishOptions,
      });

      if (!Garfish.running) {
        Garfish.run(garfishOptions);
      }
    }

    componentWillUnmount() {
      // close auto render able
      Garfish.router.setRouterConfig({ listening: false });
      logger('MApp componentWillUnmount');
    }

    render() {
      return <div id={generateSubAppContainerKey()}></div>;
    }
  }

  return Loadable(MApp)(manifest?.loadable);
}
