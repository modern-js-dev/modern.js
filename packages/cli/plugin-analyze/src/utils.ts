import fs from 'fs';
import { path, INTERNAL_DIR_ALAIS, INTERNAL_SRC_ALIAS } from '@modern-js/utils';
import type { Entrypoint } from '@modern-js/types';
import type { ImportStatement } from './generateCode';
import { FILE_SYSTEM_ROUTES_FILE_NAME } from './constants';

export const walkDirectory = (dir: string): string[] =>
  fs.readdirSync(dir).reduce<string[]>((previous, filename) => {
    const filePath = path.join(dir, filename);
    if (fs.statSync(filePath).isDirectory()) {
      return [...previous, ...walkDirectory(filePath)];
    } else {
      return [...previous, filePath];
    }
  }, []);

export const getDefaultImports = ({
  entrypoint,
  srcDirectory,
}: {
  entrypoint: Entrypoint;
  srcDirectory: string;
}): ImportStatement[] => {
  const { entryName, fileSystemRoutes, customBootstrap, entry } = entrypoint;
  const imports = [
    {
      specifiers: [{ local: 'React' }],
      value: 'react',
    },
    {
      specifiers: [{ imported: 'createApp' }, { imported: 'bootstrap' }],
      value: '@modern-js/runtime',
    },
    customBootstrap && {
      specifiers: [{ local: 'customBootstrap' }],
      value: customBootstrap.replace(srcDirectory, INTERNAL_SRC_ALIAS),
    },
  ].filter(Boolean) as ImportStatement[];

  if (fileSystemRoutes) {
    const route: ImportStatement = {
      specifiers: [{ imported: 'routes' }],
      value: `${INTERNAL_DIR_ALAIS}/${entryName}/${FILE_SYSTEM_ROUTES_FILE_NAME}`,
    };
    if (fileSystemRoutes.globalApp) {
      imports.push({
        specifiers: [{ local: 'App' }],
        value: fileSystemRoutes.globalApp.replace(
          srcDirectory,
          INTERNAL_SRC_ALIAS,
        ),
      });
    } else {
      route.initialize = 'const App = false;';
    }

    imports.push(route);
  } else {
    imports.push({
      specifiers: [{ local: 'App' }],
      value: entry.replace(srcDirectory, INTERNAL_SRC_ALIAS),
    });
  }

  return imports;
};
