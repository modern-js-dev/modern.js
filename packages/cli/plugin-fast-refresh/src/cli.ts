import { isFastRefresh } from '@modern-js/utils';
import type { CliPlugin } from '@modern-js/core';

export default (): CliPlugin => ({
  name: '@modern-js/plugin-fast-refresh',

  setup: () => ({
    config() {
      return {
        tools: {
          webpackChain: (chain, { name }) => {
            if (name === 'client' && isFastRefresh()) {
              const ReactFastRefreshPlugin = require('@pmmmwh/react-refresh-webpack-plugin');
              chain.plugin('react-fast-refresh').use(ReactFastRefreshPlugin, [
                {
                  overlay: false,
                  exclude: [/node_modules/],
                },
              ]);

              const loaders = chain.module.rule('loaders');

              const babelOptions = loaders
                .oneOf('js')
                .use('babel')
                .get('options');

              loaders
                .oneOf('js')
                .use('babel')
                .options({
                  ...babelOptions,
                  plugins: [
                    ...(babelOptions.plugins || []),
                    require.resolve('react-refresh/babel'),
                  ],
                });
            }
          },
        },
      };
    },
  }),
});
