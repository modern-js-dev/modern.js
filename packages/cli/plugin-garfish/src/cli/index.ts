import path from 'path';
import { createRuntimeExportsUtils, PLUGIN_SCHEMAS } from '@modern-js/utils';
import type { CliHookCallbacks, CliPlugin } from '@modern-js/core';
import type WebpackChain from 'webpack-chain';
import { logger } from '../util';
import {
  getRuntimeConfig,
  makeProvider,
  makeRenderFunction,
  setRuntimeConfig,
} from './utils';

export const externals = { 'react-dom': 'react-dom', react: 'react' };

type Initializer = CliPlugin['setup'];
export type LifeCycle = CliHookCallbacks;

export const initializer: Initializer = ({
  useAppContext,
  useResolvedConfigContext,
}) => {
  let pluginsExportsUtils: ReturnType<typeof createRuntimeExportsUtils>;
  let runtimeExportsUtils: ReturnType<typeof createRuntimeExportsUtils>;
  const runtimePluginName =
    process.env.PLUGIN_RUNTIME_PATH || '@modern-js/runtime/plugins';

  return {
    validateSchema() {
      return PLUGIN_SCHEMAS['@modern-js/plugin-garfish'];
    },
    resolvedConfig: async config => {
      const { resolved } = config;
      const { masterApp, router } = getRuntimeConfig(resolved);

      const nConfig = {
        resolved: {
          ...resolved,
        },
      };

      if (masterApp) {
        // basename does not exist use router's basename
        setRuntimeConfig(
          nConfig.resolved,
          'masterApp',
          Object.assign(typeof masterApp === 'object' ? { ...masterApp } : {}, {
            basename:
              router?.historyOptions?.basename || router?.basename || '/',
          }),
        );
      }

      logger(`resolvedConfig`, {
        runtime: nConfig.resolved.runtime,
        deploy: nConfig.resolved.deploy,
        server: nConfig.resolved.server,
      });
      return nConfig;
    },
    config() {
      // eslint-disable-next-line react-hooks/rules-of-hooks
      const config = useAppContext();
      // eslint-disable-next-line react-hooks/rules-of-hooks
      const resolveOptions = useResolvedConfigContext();

      pluginsExportsUtils = createRuntimeExportsUtils(
        config.internalDirectory,
        'plugins',
      );

      runtimeExportsUtils = createRuntimeExportsUtils(
        config.internalDirectory,
        'index',
      );

      return {
        source: {
          alias: {
            '@modern-js/runtime/plugins': pluginsExportsUtils.getPath(),
          },
        },
        tools: {
          webpack: (
            webpackConfig: any,
            {
              chain,
              webpack,
              env = process.env.NODE_ENV || 'development',
            }: { chain: WebpackChain; webpack: any; env: string },
          ) => {
            if (resolveOptions?.deploy?.microFrontend) {
              chain.output.libraryTarget('umd');
              chain.devServer.headers({
                'Access-Control-Allow-Origin': '*',
              });
              if (resolveOptions?.server?.port) {
                chain.output.publicPath(
                  env === 'development'
                    ? `//localhost:${resolveOptions.server.port}/`
                    : webpackConfig.output.publicPath,
                );
              }
              // add comments avoid sourcemap abnormal
              if (webpack.BannerPlugin) {
                chain
                  .plugin('banner')
                  .use(webpack.BannerPlugin, [{ banner: 'Micro front-end' }]);
              }
              const { enableHtmlEntry = true, externalBasicLibrary = false } =
                typeof resolveOptions?.deploy?.microFrontend === 'object'
                  ? resolveOptions?.deploy?.microFrontend
                  : {};
              // external
              if (externalBasicLibrary) {
                chain.externals(externals);
              }
              // use html mode
              if (!enableHtmlEntry) {
                chain.output.filename('index.js');
                chain.plugins.delete('html-main');
                chain.optimization.runtimeChunk(false);
                chain.optimization.splitChunks({
                  chunks: 'async',
                });
              }
            }
            const resolveWebpackConfig = chain.toConfig();
            logger('webpackConfig', {
              output: resolveWebpackConfig.output,
              externals: resolveWebpackConfig.externals,
              env,
              alias: resolveWebpackConfig.resolve?.alias,
            });
          },
        },
      };
    },
    addRuntimeExports() {
      const mfPackage = path.resolve(__dirname, '../../../../');
      const addExportStatement = `export { default as garfish, default as masterApp } from '${mfPackage}'`;
      logger('exportStatement', addExportStatement);
      pluginsExportsUtils.addExport(addExportStatement);

      runtimeExportsUtils.addExport(`export * from '${mfPackage}'`);
    },
    modifyEntryImports({ entrypoint, imports }: any) {
      // eslint-disable-next-line react-hooks/rules-of-hooks
      const config = useResolvedConfigContext();
      const { masterApp } = getRuntimeConfig(config);

      if (masterApp) {
        imports.push({
          value: runtimePluginName,
          specifiers: [
            {
              imported: 'garfish',
            },
          ],
        });
        imports.push({
          value: runtimePluginName,
          specifiers: [
            {
              imported: 'masterApp',
            },
          ],
        });
      }

      imports.push({
        value: 'react-dom',
        specifiers: [
          {
            imported: 'unmountComponentAtNode',
          },
          {
            imported: 'createPortal',
          },
        ],
      });

      return { imports, entrypoint };
    },
    modifyEntryRuntimePlugins({ entrypoint, plugins }: any) {
      // eslint-disable-next-line react-hooks/rules-of-hooks
      const config = useResolvedConfigContext();
      const { masterApp } = getRuntimeConfig(config);

      if (masterApp) {
        logger('garfishPlugin options', masterApp);

        plugins.push({
          name: 'garfish',
          args: 'masterApp',
          options:
            masterApp === true ? JSON.stringify({}) : JSON.stringify(masterApp),
        });
      }
      return { entrypoint, plugins };
    },
    modifyEntryRenderFunction({ entrypoint, code }: any) {
      // eslint-disable-next-line react-hooks/rules-of-hooks
      const config = useResolvedConfigContext();

      if (!config?.deploy?.microFrontend) {
        return { entrypoint, code };
      }
      const nCode = makeRenderFunction(code);
      logger('makeRenderFunction', nCode);
      return {
        entrypoint,
        code: nCode,
      };
    },
    modifyEntryExport({ entrypoint, exportStatement }: any) {
      // eslint-disable-next-line react-hooks/rules-of-hooks
      const config = useResolvedConfigContext();
      if (config?.deploy?.microFrontend) {
        const exportStatementCode = makeProvider();
        logger('exportStatement', exportStatementCode);
        return {
          entrypoint,
          exportStatement: exportStatementCode,
        };
      }

      return {
        entrypoint,
        exportStatement,
      };
    },
  };
};

export default (): CliPlugin => ({
  name: '@modern-js/plugin-garfish',
  setup: initializer,
});
