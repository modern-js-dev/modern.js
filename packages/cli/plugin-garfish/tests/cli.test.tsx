import '@testing-library/jest-dom';
import { manager } from '@modern-js/core';
import WebpackChain from 'webpack-chain';
import GarfishPlugin, { externals, resolvedConfig } from '../src/cli';
import { getRuntimeConfig, makeRenderFunction, setRuntimeConfig } from '../src/cli/utils';

describe('plugin-garfish cli', () => {
  test('cli garfish basename', async () => {
    expect(GarfishPlugin().name).toBe('@modern-js/plugin-garfish');
    const basename = '/test';
    const resolveConfig = {
      resolved: {
        runtime: {
          router: {
            historyOptions: { basename }
          },
          masterApp: {}
        },
      }
    };
    const config = await resolvedConfig(resolveConfig as any);
    expect(config.resolved.runtime.masterApp.basename).toBe(basename);
  });

  test('cli makeRender function', ()=>{
    const code = `
      const router = (config)=> config;
      const App = {};
      const routerConfig = router({...App?.config?.router, ...App?.config?.features?.router});
      const resultConfig = {
        routerConfig,
        renderByProvider: true,
      };

      if (IS_BROWSER) {
        resultConfig.renderByProvider = false;
      }
      return resultConfig;
    `;
    const generateNewRenderFn = new Function('appInfo', 'IS_BROWSER', '__GARFISH_EXPORTS__', makeRenderFunction(code));

    // render byGarfish but don't provider appInfo
    expect(generateNewRenderFn(undefined, true, false)).toBe(null);

    // run alone
    expect(generateNewRenderFn(undefined, true)).toMatchObject({
      renderByProvider: false
    });

    // render ByGarfish and provider appInfo
    expect(generateNewRenderFn({ basename: '/test' }, true, true)).toMatchObject({
      renderByProvider: true,
      routerConfig: {
        basename: '/test',
        historyOptions: {
          basename: '/test'
        }
      }
    });
  });

  test('cli get runtime config', ()=>{
    const runtimeConfig = getRuntimeConfig({
      runtime: {
        masterApp: {
          basename: '/test'
        }
      }
    });
    expect(runtimeConfig).toMatchObject({
      masterApp: {
        basename: '/test'
      }
    });
  });

  test('cli get runtime features config', ()=>{
    const runtimeConfig = getRuntimeConfig({
      runtime: {
        masterApp: {
          basename: '/test'
        },
        features: {
          masterApp: {
            basename: '/test2'
          }
        }
      }
    });

    expect(runtimeConfig).toMatchObject({
      masterApp: {
        basename: '/test2'
      }
    });
  });

  test('cli set runtime config', ()=>{
    const runtimeConfig = {
      runtime: {
        masterApp: {
          basename: '/test'
        }
      }
    };

    setRuntimeConfig(runtimeConfig, 'masterApp', true);

    expect(runtimeConfig.runtime.masterApp).toBe(true);
  });

  test('cli set runtime features config', ()=>{
    const runtimeConfig = {
      runtime: {
        features: {
          masterApp: {
            basename: '/test'
          }
        }
      }
    };

    setRuntimeConfig(runtimeConfig, 'masterApp', true);

    expect(runtimeConfig.runtime.features.masterApp).toBe(true);
  });

  test('webpack config close external and use js entry', async ()=>{
    const resolveConfig: any = {
      deploy: {
        microFrontend: {
          externalBasicLibrary: true,
          enableHtmlEntry: false,
        },
      },
      server: {
        port: 8080
      }
    };

    const main = manager.clone({
      useResolvedConfigContext: ()=>resolveConfig
    }).usePlugin(GarfishPlugin);

    const runner = await main.init();
    await runner.prepare();
    const config: any = await runner.config();
    const webpackConfig = new WebpackChain();

    config[0].tools.webpack({}, {
      chain: webpackConfig,
      webpack: jest.fn(),
      env: 'development'
    });

    const generateConfig = webpackConfig.toConfig();
    expect(generateConfig).toMatchObject({
      output: {
        libraryTarget: 'umd',
        publicPath: '//localhost:8080/',
        filename: 'index.js'
      },
      externals,
      optimization: { runtimeChunk: false, splitChunks: { chunks: 'async' } }
    });
  })

  test('webpack config default micro config', async ()=>{
    const resolveConfig: any = {
      deploy: {
        microFrontend: true,
      },
      server: {
        port: '8080'
      }
    };

    const main = manager.clone({
      useResolvedConfigContext: ()=>resolveConfig
    }).usePlugin(GarfishPlugin);
    const runner = await main.init();
    await runner.prepare();
    const config: any = await runner.config();
    const webpackConfig = new WebpackChain();

    config[0].tools.webpack({}, {
      chain: webpackConfig,
      webpack: jest.fn(),
      env: 'development'
    });

    const generateConfig = webpackConfig.toConfig();
    expect(generateConfig).toMatchObject({
      output: {
        libraryTarget: 'umd',
        publicPath: '//localhost:8080/'
      }
    });
    expect(generateConfig.externals).toBeUndefined();
    expect(generateConfig.output.filename).toBeUndefined();
  });
});
