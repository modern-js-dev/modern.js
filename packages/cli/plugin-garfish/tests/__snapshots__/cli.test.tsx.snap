// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin-garfish cli cli get runtime config 1`] = `
Object {
  "masterApp": Object {
    "basename": "/test",
  },
}
`;

exports[`plugin-garfish cli cli get runtime features config 1`] = `
Object {
  "masterApp": Object {
    "basename": "/test2",
  },
}
`;

exports[`plugin-garfish cli cli makeRender function 1`] = `
Object {
  "renderByProvider": true,
  "routerConfig": Object {
    "basename": "/test",
    "historyOptions": Object {
      "basename": "/test",
    },
  },
}
`;

exports[`plugin-garfish cli cli set runtime config 1`] = `
Object {
  "masterApp": true,
}
`;

exports[`plugin-garfish cli cli set runtime features config 1`] = `
Object {
  "features": Object {
    "masterApp": true,
  },
}
`;

exports[`plugin-garfish cli webpack config close external and use js entry 1`] = `
Object {
  "devServer": Object {
    "headers": Object {
      "Access-Control-Allow-Origin": "*",
    },
  },
  "externals": Object {
    "react": "react",
    "react-dom": "react-dom",
  },
  "optimization": Object {
    "runtimeChunk": false,
    "splitChunks": Object {
      "chunks": "async",
    },
  },
  "output": Object {
    "filename": "index.js",
    "libraryTarget": "umd",
    "publicPath": "//localhost:8080/",
  },
}
`;

exports[`plugin-garfish cli webpack config default micro config 1`] = `
Object {
  "devServer": Object {
    "headers": Object {
      "Access-Control-Allow-Origin": "*",
    },
  },
  "output": Object {
    "libraryTarget": "umd",
    "publicPath": "//localhost:8080/",
  },
}
`;
